google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -1330,6 +1331,166 @@ private Map<String, Object> getData() throws ExecutionException, InterruptedExce
     return randomDoc.get().get().getData();
   }
 
+  @Test
+  public void writeAndReadVectorEmbeddings() throws ExecutionException, InterruptedException {
+    Map<String, VectorValue> expected = new HashMap<>();
+
+    randomDoc
+        .create(
+            map(
+                "vector0",
+                FieldValue.vector(new double[] {0.0}),
+                "vector1",
+                FieldValue.vector(new double[] {1, 2, 3.99})))
+        .get();
+    randomDoc
+        .set(
+            map(
+                "vector0",
+                FieldValue.vector(new double[] {0.0}),
+                "vector1",
+                FieldValue.vector(new double[] {1, 2, 3.99}),
+                "vector2",
+                FieldValue.vector(new double[] {0, 0, 0})))
+        .get();
+    randomDoc.update(map("vector3", FieldValue.vector(new double[] {-1, -200, -9999}))).get();
+
+    expected.put("vector0", FieldValue.vector(new double[] {0.0}));
+    expected.put("vector1", FieldValue.vector(new double[] {1, 2, 3.99}));
+    expected.put("vector2", FieldValue.vector(new double[] {0, 0, 0}));
+    expected.put("vector3", FieldValue.vector(new double[] {-1, -200, -9999}));
+
+    Map<String, Object> actual = getData();
+
+    assertTrue(actual.get("vector0") instanceof VectorValue);
+    assertTrue(actual.get("vector1") instanceof VectorValue);
+    assertTrue(actual.get("vector2") instanceof VectorValue);
+    assertTrue(actual.get("vector3") instanceof VectorValue);
+
+    assertArrayEquals(
+        expected.get("vector0").getArray(),
+        ((VectorValue) actual.get("vector0")).getArray(),
+        DOUBLE_EPSILON);
+    assertArrayEquals(
+        expected.get("vector1").getArray(),
+        ((VectorValue) actual.get("vector1")).getArray(),
+        DOUBLE_EPSILON);
+    assertArrayEquals(
+        expected.get("vector2").getArray(),
+        ((VectorValue) actual.get("vector2")).getArray(),
+        DOUBLE_EPSILON);
+    assertArrayEquals(
+        expected.get("vector3").getArray(),
+        ((VectorValue) actual.get("vector3")).getArray(),
+        DOUBLE_EPSILON);
+  }
+
+  @Test
+  public void listenToDocumentsWithVectors() throws Exception {
+    final Semaphore semaphore = new Semaphore(0);
+    ListenerRegistration registration = null;
+    DocumentReference ref = randomColl.document();
+    AtomicReference<String> failureMessage = new AtomicReference(null);
+    int totalPermits = 5;
+
+    try {
+      registration =
+          randomColl
+              .whereEqualTo("purpose", "vector tests")
+              .addSnapshotListener(
+                  (value, error) -> {
+                    try {
+                      DocumentSnapshot docSnap =
+                          value.isEmpty() ? null : value.getDocuments().get(0);
+
+                      switch (semaphore.availablePermits()) {
+                        case 0:
+                          assertNull(docSnap);
+                          ref.create(
```

Comment:
I think if `create` fails, it will do so silently until your test times out. Maybe you should make this blocking by adding `.get()`, that way problems will throw exception that can be used to cause test to fail rather than timeout.

Reply:
Good insight. Updated this.